### PR TITLE
Fix 966

### LINKED ISSUE #966: Database migrations fail to add foreign key
I have an integer column called `orderId`.

Now I want to add a foreign key to it:

```javascript
module.exports = {
  up: function(migration, DataTypes, done) {
    migration.changeColumn("orderitems", "orderId", 
      {
        type: DataTypes.INTEGER,
        references: "orders",
        referenceKey: "id",
        onUpdate: "CASCADE",
        onDelete: "RESTRICT"
      }
    ).complete(done);
  }
}
```

but the migration fails with this:
```
{ '0':
   { [error: syntax error at or near "REFERENCES"]
     length: 93,
     name: 'error',
     severity: 'ERROR',
     code: '42601',
     detail: undefined,
     hint: undefined,
     position: '185',
     internalPosition: undefined,
     internalQuery: undefined,
     where: undefined,
     file: 'scan.l',
     line: '1002',
     routine: 'scanner_yyerror' } }
Completed in 9ms

events.js:74
        throw TypeError('Uncaught, unspecified "error" event.');
              ^
TypeError: Uncaught, unspecified "error" event.
    at TypeError (<anonymous>)
    at EventEmitter.emit (events.js:74:15)
    at null.<anonymous> (/Users/hoitz/develop/salad/node_modules/sequelize/lib/migrator.js:95:44)
    at EventEmitter.emit (events.js:98:17)
    at module.exports.finish (/Users/hoitz/develop/salad/node_modules/sequelize/lib/query-chainer.js:138:30)
    at exec (/Users/hoitz/develop/salad/node_modules/sequelize/lib/query-chainer.js:92:16)
    at onError (/Users/hoitz/develop/salad/node_modules/sequelize/lib/query-chainer.js:72:11)
    at EventEmitter.emit (events.js:95:17)
    at /Users/hoitz/develop/salad/node_modules/sequelize/lib/migration.js:65:19
    at null.<anonymous> (/Users/hoitz/develop/salad/node_modules/sequelize/lib/emitters/custom-event-emitter.js:52:38)
````

The error is caused by this query:

```
ALTER TABLE "orderitems" 
ALTER COLUMN "orderId" TYPE INTEGER REFERENCES "orders" ("id") 
ON DELETE RESTRICT 
ON UPDATE CASCADE;
````

which is invalid SQL. It should add a constraint to the column.

### PR DESCRIPTION
**Pull request by sushantdhiman**
Fix #966 

I have added support for `mysql`,  `mssql`, `postgres`, `mariadb`. `changeColumn()` will be able to add the `foreign keys` now.

`sqlite` doesn't support alerting columns natively.

Also once merged it will close #5014

### CODE REVIEW

**Comment by sushantdhiman** [comment]
cc @janmeier , @mickhansen

File: test/integration/query-interface.test.js [hunk 1]
```diff
@@ -123,20 +123,20 @@ describe(Support.getTestDialectTeaser('QueryInterface'), function() {
...
-              tableName: 'table'
-            }, {logging: log}).then(function(indexes) {
-              expect(indexes.length).to.eq(1);
-              count = 0;
-            });
+          }, {logging: log}).then(function(indexes) {
+            expect(indexes.length).to.eq(1);
+            count = 0;
           });
+        });
```
Comment (mickhansen) [inline test/integration/query-interface.test.js]: What changed here exactly?

File: test/integration/query-interface.test.js [hunk 2]
```diff
@@ -2,18 +2,18 @@
...
+, DataTypes = require(__dirname + '/../../lib/data-types')
+, dialect = Support.getTestDialect()
+, _ = require('lodash')
+, count = 0
+, log = function (sql) {
+  // sqlite fires a lot more querys than the other dbs. this is just a simple hack, since i'm lazy
+  if (dialect !== 'sqlite' || count === 0) {
+    count++;
+  }
+};
```
Comment (mickhansen) [inline test/integration/query-interface.test.js]: Why are you removing indentation here?

**Comment by mickhansen** [comment]
I didn't think dialects supported foreign key changes as part of a `ALTER TABLE CHANGE COLUMN` call actually.

Would love to see some sql unit tests on this.

File: test/integration/query-interface.test.js (hunk 2 above)
Comment (sushantdhiman) [inline test/integration/query-interface.test.js]: Ah auto indent , will fix

File: test/integration/query-interface.test.js (hunk 1 above)
Comment (sushantdhiman) [inline test/integration/query-interface.test.js]: CRLF

File: test/integration/query-interface.test.js [hunk 3]
```diff
@@ -174,46 +174,46 @@ describe(Support.getTestDialectTeaser('QueryInterface'), function() {
...
 
           switch (dialect) {
             case 'sqlite':
-              expect(username.defaultValue).to.be.undefined;
-              break;
+            expect(username.defaultValue).to.be.undefined;
+            break;
             default:
-              expect(username.defaultValue).to.be.null;
+            expect(username.defaultValue).to.be.null;
```
Comment (mickhansen) [inline test/integration/query-interface.test.js]: weird identation changes

File: test/integration/query-interface.test.js (hunk 1 above)
Comment (mickhansen) [inline test/integration/query-interface.test.js]: best to keep style changes out of feature changes, easier to review and reason about.
You can then submit separate style changes pull requests, or atleast separate commits :)

File: test/integration/query-interface.test.js (hunk 3 above)
Comment (sushantdhiman) [inline test/integration/query-interface.test.js]: Yeah, sorry will change

File: test/integration/query-interface.test.js (hunk 1 above)
Comment (sushantdhiman) [inline test/integration/query-interface.test.js]: I didn't want to confuse by adding these random changes either, Atom is to blame. Will remove them

**Comment by sushantdhiman** [comment]
@mickhansen I have added the `changeColumn` unit tests.

File: test/unit/sql/change-column.test.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                   key:   'id'
+                 },
+                 onUpdate: 'cascade',
+                 onDelete: 'cascade'
+               }).then(function(sql){
+                 expectsql(sql, {
+                   mssql: 'ALTER TABLE [users]  ADD CONSTRAINT [level_id_foreign_idx] FOREIGN KEY ([level_id]) REFERENCES [level] ([id]) ON DELETE CASCADE;',
+                   mysql: 'ALTER TABLE `users`  ADD CONSTRAINT `level_id_foreign_idx` FOREIGN KEY (`level_id`) REFERENCES `level` (`id`) ON DELETE CASCADE ON UPDATE CASCADE;',
+                   //Postgres issue multiple queries when changing schema to drop default indexes
+                   postgres: 'ALTER TABLE "users" ALTER COLUMN "level_id" DROP NOT NULL;ALTER TABLE "users" ALTER COLUMN "level_id" DROP DEFAULT;ALTER TABLE "users"  ADD CONSTRAINT "level_id_foreign_idx" FOREIGN KEY ("level_id") REFERENCES "level" ("id") ON DELETE CASCADE ON UPDATE CASCADE;',
```
Comment (mickhansen) [inline test/unit/sql/change-column.test.js]: I know you likely just tested against what it's doing, but it's really sort of a bug, it should not be dropping NOT NULL or the default value here.

File: lib/dialects/mssql/query-generator.js [hunk 5]
```diff
@@ -181,13 +181,18 @@ var QueryGenerator = {
...
+      if (definition.match(/REFERENCES/)) {
+        query = query.replace('ALTER COLUMN', '');
+        definition = definition.replace(/.+?(?=REFERENCES)/,'');
+        attrString.push('ADD CONSTRAINT ' + this.quoteIdentifier(attributeName + '_foreign_idx') + ' FOREIGN KEY (' + this.quoteIdentifier(attributeName) + ') ' + definition);
+      } else {
+        attrString.push(Utils._.template('<%= attrName %> <%= definition %>')({
+          attrName: this.quoteIdentifier(attributeName),
+          definition: definition
+        }));
+      }
```
Comment (mickhansen) [inline lib/dialects/mssql/query-generator.js]: Sort of a hacky fix, you are looking for incorrect SQL and then fixing it, perhaps we should fix it more generally.

Comment (sushantdhiman) [inline lib/dialects/mssql/query-generator.js]: I have changed the logic. Now code will not override the query template

**Comment by sushantdhiman** [comment]
@mickhansen I have changed the logic for generating queries. Please review

File: lib/dialects/mssql/query-generator.js (hunk 5 above)
Comment (mickhansen) [inline lib/dialects/mssql/query-generator.js]: Right but it's still matching REFERENCES against the SQL.
This syntax will not support both changing a columns default value and setting a constraint will it?

Comment (sushantdhiman) [inline lib/dialects/mssql/query-generator.js]: Now they will support adding constraint and change in single step

**Comment by sushantdhiman** [comment]
Constraint can be added and changed at same time now, Let me know if there are other changes in PR

File: lib/dialects/mssql/query-generator.js (hunk 5 above)
Comment (mickhansen) [inline lib/dialects/mssql/query-generator.js]: Still not totally happy with `if (definition.match(/REFERENCES/)) {`

Comment (sushantdhiman) [inline lib/dialects/mssql/query-generator.js]: @mickhansen Can you suggest other way to detect references in query. (All dialects are doing this)

Comment (mickhansen) [inline lib/dialects/mssql/query-generator.js]: Can you link to where else were doing that? Will want to tag that for refactor then.

File: test/unit/sql/change-column.test.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                 },
+                 onUpdate: 'cascade',
+                 onDelete: 'cascade'
+               }).then(function(sql){
+                 expectsql(sql, {
+                   mssql: 'ALTER TABLE [users] ADD CONSTRAINT [level_id_foreign_idx] FOREIGN KEY ([level_id]) REFERENCES [level] ([id]) ON DELETE CASCADE;',
+                   mysql: 'ALTER TABLE `users` ADD CONSTRAINT `level_id_foreign_idx` FOREIGN KEY (`level_id`) REFERENCES `level` (`id`) ON DELETE CASCADE ON UPDATE CASCADE;',
+                   postgres: 'ALTER TABLE "users"  ADD CONSTRAINT "level_id_foreign_idx" FOREIGN KEY ("level_id") REFERENCES "level" ("id") ON DELETE CASCADE ON UPDATE CASCADE;',
+                 });
+               });
```
Comment (mickhansen) [inline test/unit/sql/change-column.test.js]: Indent issues in this test file

File: lib/dialects/mssql/query-generator.js (hunk 5 above)
Comment (sushantdhiman) [inline lib/dialects/mssql/query-generator.js]: https://github.com/sequelize/sequelize/blob/master/lib/dialects/postgres/query-generator.js#L231
https://github.com/sequelize/sequelize/blob/master/lib/dialects/postgres/query-generator.js#L251

Comment (mickhansen) [inline lib/dialects/mssql/query-generator.js]: Ah yeah that's not great either. But we'll move forward in the spirit of getting it fixed :)

**Comment by mickhansen** [comment]
Needs a rebase and an indentation fix and then i'll merge

**Comment by sushantdhiman** [comment]
@mickhansen Fixed indentation and rebased

**Comment by mickhansen** [comment]
Thanks again for your hard work and patience @sushantdhiman

**Comment by sushantdhiman** [comment]
Thanks @mickhansen

**Comment by jocull** [comment]
Thanks so much! Will we know when this hits npm?

**Comment by mickhansen** [comment]
@jocull We track releases on the github releases tab

**Comment by jocull** [comment]
Thank you!

**Comment by rayshan** [comment]
Thanks @sushantdhiman & team!

**Pull request merged**